kafka/conn.py

```diff
@@ -772,9 +778,14 @@ def _send(self, request, blocking=True):
 
         log.debug('%s Request %d: %s', self, correlation_id, request)
         if request.expect_response():
-            sent_time = time.time()
-            assert correlation_id not in self.in_flight_requests, 'Correlation ID already in-flight!'
-            self.in_flight_requests[correlation_id] = (future, sent_time)
+            with self._ifr_lock:
+                if self.disconnected():
+                    log.warning("%s: Race condition: connection already closed.")
```

Comment:
There are actually 2 problems:
1. Connection is closed with "Protocol out of sync". I suspect I know the root cause but I am intentionally not fixing it now. When I deploy this version, this log message will tell me whether I'm right about it and I will know for sure. That's when I fix it.
2. When connection is closed, it leaves in_flight_requests hanging.
The timeline is as follows:
- send() is called, sends a message
- before future is saved to `in_flight_requests`, another thread calls `close()` (such as because of problem number 1)
- `close()` terminates all futures in `in_flight_requests`, but there are none there
- thread which called `send()` stores the future to disconnected connection's `in_flight_requests`
- the future remains unresolved forever = worker deadlocks

Reply:
I will deploy this on production and check the behavior. If my theory is right, I will eventually get to see this sequence of events:

1. "protocol out of sync error"
2. Connection close
3. Race condition warning
4. Worker reestabilishing connection successfully